RDDRONE-FMUK66 enable i2c and spi sensors.

sumitbatra-nxp opened the pull request:
This PR includes code to add following functionality on rddrone_fmuk66 board -
1. Makes 3.3V onboard power supply for sensor enablement to be regulated by the zephyr regulator driver.
2. Resolves issues related to working of I2C based sensors on internal I2C bus.
3. Adds support for fxos8700 and fxas21002 sensors over SPI bus.  
4. Adds rgb leds for indicating pwm outputs

danieldegrasse commented on boards/arm/rddrone_fmuk66/board.c:
```diff
@@ -0,0 +1,49 @@
+static int setup()
+{
+
+	if (IS_ENABLED(CONFIG_SENSOR)) {
```
Are these delays specific to the RDDRONE? If not, I would rather see them integrated into the sensor driver. This could be done with some sort of KConfig, which enabled the sensor driver to delay at startup so the sensor could power up.

sumitbatra-nxp replied:
These changes are specific to rddrone fmuk66. 
There is a 3.3V supply for several onboard sensor, which is enabled through a GPIO pin.
Changes in this file make sure that the sequence of enablement of GPIO module, regulator driver, this board specific module and the sensor modules are in a particular sequence and have a small delay between their respective enablement

danieldegrasse replied:
Thanks for the response, I'm fine leaving this delay where it currently is in that case.

danieldegrasse commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -66,17 +70,61 @@
+	servo: servo {
+		compatible = "pwm-servo";
+		pwms = <&ftm0 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>, // FMU_CH1
+			   <&ftm0 3 PWM_MSEC(20) PWM_POLARITY_NORMAL>, // FMU_CH2
```
Comments must use C89 style (`/* */`)

danieldegrasse commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -280,6 +346,37 @@ zephyr_udc0: &usbotg {
 	status = "okay";
 	pinctrl-0 = <&spi1_default>;
 	pinctrl-names = "default";
+        /* cs-gpios needs to be populated as per the chip select
```
Format comments like this:

```
/* My comment!
 * this comment has useful info
 */
```

sumitbatra-nxp replied:
Acknowledged

MaureenHelm commented:
> * Makes 3.3V onboard power supply for sensor enablement to be regulated by the zephyr regulator driver.
> * Resolves issues related to working of I2C based sensors on internal I2C bus.
> * Adds support for fxos8700 and fxas21002 sensors over SPI bus.

Please split up these changes into separate commits

danieldegrasse reviewed (commented):
Thank you for the contribution, and your work on supporting this board.

A few comments here:
- You need to split these changes into separate commits. From your comment, I'd suggest something like the following:
    - enable 3.3V regulator lines for this board
    - resolve issues with I2C sensors on I2C bus
    - add support for fxos8700 on SPI bus
    - add support for fxas21002 on SPI bus
- Commits need to have more detail, and appropriate titles. Take a look here
for guidelines: https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines In particular, commits need a body to be included.

gmarull commented on boards/arm/rddrone_fmuk66/board.c:
```diff
@@ -0,0 +1,49 @@
+static int setup()
+{
+
+	if (IS_ENABLED(CONFIG_SENSOR)) {
+		/* Initialization chain of rddrone fmuk66 board requires some delays before on board
+		 * sensors could be accessed after power up. In particular bme680 and bmm150
+		 * sensors require, 2ms and 1ms power on delay respectively. In order not to sum
+		 * delays, common delay is introduced in the board start up file. This code is
+		 * executed after sensors are powered up and before their initialization.
+		 * It's ensured by build asserts at the beginning of this file.
+		 */
+		k_msleep(2);
+	}
```
this seems like a workaround for bus voltage stabilization, why not use `startup-delay-us` in the regulator?

sumitbatra-nxp replied:
Will revert on this very shortly

sumitbatra-nxp replied:
@gmarull I have verified that startup-delay-us does not resolve the i2c sensor issue on rddrone board
I can further compare the effect of the delay I added vs startup-delay-us later, but for now please accept this change.

MaureenHelm replied:
Why doesn't `startup-delay-us` work? Isn't that what it's for?

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66-pinctrl.dtsi:
```diff
@@ -76,20 +76,30 @@
 			drive-strength = "low";
 			slew-rate = "fast";
 		};
+                group1 {
```
please check your editor settings, this mixes spaces/tabs

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -66,17 +70,61 @@
+	/* This regulator enables the 3V3_S line, which powers sensors on-board. */
+	reg-3v3-s {
+		compatible = "regulator-fixed";
+		label = "reg-3v3-s";
```
label is deprecated, please remove

sumitbatra-nxp replied:
Acknowledged

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -266,6 +314,24 @@ zephyr_udc0: &usbotg {
+		reg = <0x76>;
+	};
+
+        bmm150: bmm150@10 {
```
same, please fix your editor settings

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66_defconfig:
```diff
@@ -13,3 +14,9 @@ CONFIG_OSC_EXTERNAL=y
+
+# Enable Regulators
+CONFIG_REGULATOR=y
+CONFIG_REGULATOR_FIXED=y
```
this is redundant

gmarull replied:
not addressed

bperseghetti replied:
@gmarull isn't CONFIG_REGULATOR_FIXED dependent on CONFIG_REGULATOR when doing manual config? Is it not also required in defconfig?

Are you suggesting to remove all: CONFIG_REGULATOR_XXX from the defconfig?

gmarull replied:
`CONFIG_REGULATOR_FIXED` is automatically enabled if `CONFIG_REGULATOR=y` and a node in DT with compatible `regulator-fixed` has its status set to `okay`.

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66_defconfig:
```diff
@@ -13,3 +14,9 @@ CONFIG_OSC_EXTERNAL=y
+CONFIG_REGULATOR=y
+CONFIG_REGULATOR_FIXED=y
+CONFIG_REGULATOR_FIXED_INIT_PRIORITY=75
+CONFIG_REGULATOR_LOG_LEVEL_WRN=y
```
please stick to minimal options in board files, per porting guidelines. enabling log levels here is wrong

sumitbatra-nxp replied:
As per your suggestion, I have removed - CONFIG_REGULATOR_LOG_LEVEL_WRN
But rest is necessary for most of the sensors on this board to work.
So, I would request you to accept the others.

sumitbatra-nxp commented on boards/arm/rddrone_fmuk66/board.c:
```diff
@@ -0,0 +1,49 @@
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/logging/log.h>
```
Acknowledged

danieldegrasse commented on dts/bindings/sensor/nxp,fxas21002.yaml:
```diff
@@ -5,8 +5,7 @@ description: FXAS21002 3-axis gyroscope sensor
 
-include: i2c-device.yaml
-include: spi-device.yaml
+include: [sensor-device.yaml, i2c-device.yaml, spi-device.yaml]
```
Please don't include merges in your commit history. When you push an update, please rebase your changes onto the latest version of `main` and resolve conflicts between your changes and `main` within the commit that introduces the conflicts.

bperseghetti replied:
Thanks, we rebased and fixed the conflicts in their respective commits.

bperseghetti commented:
@danieldegrasse do you have any insight on to why the twister-build-prep is failing:

```
INFO: Error found: samples/sensor/fxos8700/sample.sensor.fxos8700.hybrid on frdm_k64f (No hardware support but is one of the integration platforms)
INFO: Error found: samples/sensor/fxos8700/sample.sensor.fxos8700.accel on bbc_microbit (No hardware support but is one of the integration platforms)
INFO: Error found: samples/sensor/magn_polling/sample.sensor.magn_polling on stm32f411e_disco (No hardware support but is one of the integration platforms)
```

danieldegrasse commented:
> @danieldegrasse do you have any insight on to why the twister-build-prep is failing:

@bperseghetti Take a look at [line 4](https://github.com/zephyrproject-rtos/zephyr/blob/737405e593fd23c4f2ff1cabe9a2a606f0b343dc/samples/sensor/fxos8700/sample.yaml#L4) of the `sample.yaml` for the fxos8700:

```
depends_on: i2c gpio spi fxos8700
```

Note that `fxos8700` is absent from the `supported` list in [frdm_k64f.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/737405e593fd23c4f2ff1cabe9a2a606f0b343dc/boards/arm/frdm_k64f/frdm_k64f.yaml):

```
supported:
  - adc
  - arduino_gpio
  - arduino_i2c
  - arduino_serial
  - arduino_spi
  - can
  - counter
  - dac
  - dma
  - gpio
  - i2c
  - netif:eth
  - nvs
  - pwm
  - spi
  - usb_device
  - watchdog
 ```

I would guess this is what's causing your failure. The `depends_on` property is compared with a board's `supported` list by twister. Twister cannot prepare a test plan for these tests because the integration platform (`frdm_k64f`) does not appear to support the sample. I would drop `fxos8700` from your `depends_on` property for these samples.

danieldegrasse commented:
As an aside, your commits still need to be updated to follow [Zephyr's commit guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines):

> The subject line must include a prefix that identifies the subsystem being changed, followed by a colon, and a short title, for example: doc: update wiki references to new site


Your code also needs to utilize tabs for indentation. To speed up fixing these compliance issues, you can run checkpatch locally on a Linux machine (or via WSL) with `${ZEPHYR_BASE}/scripts/checkpatch.pl`

bperseghetti reviewed (commented):
I think this has all the changes in it now and also fixed the twister issues.

danieldegrasse reviewed (commented):
Regarding your commit messages- can you change from `boards:` to `boards:rddrone_fmuk66` to clarify which board is being updated here?

bperseghetti commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -338,6 +339,26 @@ zephyr_udc0: &usbotg {
 	status = "okay";
 	pinctrl-0 = <&spi1_default>;
 	pinctrl-names = "default";
+		/*
```
Ok, odd perl script didn't catch that:

```
perl scripts/checkpatch.pl -f boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts
boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts total: 0 errors, 0 warnings, 390 lines checked

boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts has no obvious style problems and is ready for submission.
```

bperseghetti reviewed (commented):
@danieldegrasse I rebased with commit name changes as well as fixing the indent level of the aforementioned comment block.

MaureenHelm commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66-pinctrl.dtsi:
```diff
@@ -53,7 +53,7 @@
 		group0 {
 			pinmux = <FTM0_CH0_PTC1>,
-				<FTM0_CH1_PTA4>,
+				<FTM0_CH3_PTA6>,
```
Is this change related to the regulator? It doesn't seem to belong in commit `boards:rddrone_fmuk66 Enable 3.3v regulator`

MaureenHelm commented on boards/arm/rddrone_fmuk66/Kconfig.defconfig:
```diff
@@ -20,6 +20,9 @@ config MCG_VDIV0
 config MCG_FCRDIV
 	default 1
 
+config I2C
+	default y
+
```
This change shouldn't be necessary now that sensor drivers select I2C

sumitbatra-nxp replied:
Acknowledged

MaureenHelm commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -307,6 +308,22 @@ zephyr_udc0: &usbotg {
 	status = "okay";
 	pinctrl-0 = <&i2c1_default>;
 	pinctrl-names = "default";
+
+	/*
+	 * This board does not have a BME280, it has a BMP280.
+	 * The two parts are incredibly similar, so the BME280 driver should work.
+	 */
+	bme280@76 {
+		compatible = "bosch,bme280";
+		status = "okay";
+		reg = <0x76>;
+	};
+
+	bmm150: bmm150@10 {
+		compatible = "bosch,bmm150";
+		status = "okay";
+		reg = <0x10>;
+	};
```
These changes don't belong in a commit that's fixing I2C for the board. Please move to a separate commit

sumitbatra-nxp replied:
@MaureenHelm .. 
Why doesn't startup-delay-us work? Isn't that what it's for?
>> Your comment makes sense, but startup-delay-us did not work for me when I tried a few days ago, but today it is working, I am testing with more thoroughly with startup-delay-us

sumitbatra-nxp replied:
Changed as per your suggestion

MaureenHelm commented on samples/sensor/magn_polling/sample.yaml:
```diff
@@ -11,3 +11,4 @@ tests:
       - sensortile_box          # lis2mdl
       - stm32f411e_disco        # lsm303agr_magn
       - stm32f3_disco           # lsm303dlhc_magn
+      - rddrone_fmuk66          # bmm150
```
We already have an integration platform for this sensor, so we don't need to add another one

sumitbatra-nxp replied:
Acknowledged

MaureenHelm commented on boards/arm/rddrone_fmuk66/Kconfig.defconfig:
```diff
@@ -23,6 +23,9 @@ config MCG_FCRDIV
 config I2C
 	default y
 
+config SPI
+	default y
+
```
This change shouldn't be necessary now that sensor drivers select SPI when they need it

sumitbatra-nxp replied:
Acknowledged

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -18,6 +18,130 @@ LOG_MODULE_REGISTER(FXOS8700, CONFIG_SENSOR_LOG_LEVEL);
+#define DIR_WRITE(a) ((a) | (1 << 7))
+#define ADDR_7(a) ((a) & (1 << 7))
+
+int fxos8700_transceive(const struct device *dev,
```
Please split the driver changes into a standalone commit

sumitbatra-nxp replied:
We have created one commit for fxos8700 sensor driver and its corresponding dts node, 
and similarly another commit for fxas21002 sensor driver and its respective dts node.

What change do you expect in this commit method ?

sumitbatra-nxp replied:
Changed as per your suggestion

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -388,8 +511,14 @@ static int fxos8700_init(const struct device *dev)
 	struct fxos8700_data *data = dev->data;
 	struct sensor_value odr = {.val1 = 6, .val2 = 250000};
 
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
```
This logic isn't quite right. We need to test if *this* device instance is on an I2C bus, not if *any* device instance is on an I2C bus.

sumitbatra-nxp replied:
Acknowledged

sumitbatra-nxp replied:
@MaureenHelm .. Acknowledged

sumitbatra-nxp replied:
Changed as per your suggestion, please check now

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.h:
```diff
@@ -5,7 +5,11 @@
  */
 
 #include <zephyr/drivers/sensor.h>
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+#include <zephyr/drivers/spi.h>
+#elif DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
 #include <zephyr/drivers/i2c.h>
+#endif
```
```suggestion
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
#include <zephyr/drivers/spi.h>
#endif

#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
#include <zephyr/drivers/i2c.h>
#endif
```

sumitbatra-nxp replied:
Acknowledged

sumitbatra-nxp replied:
@MaureenHelm .. 
Though I understand your comments for removing the #elif but fyi..
We got inspired from following sensor drivers, where similar #if, #elif constructs are used in a similar scenario-
- drivers/sensor/iis2dh/iis2dh.c
- drivers/sensor/lis2dw12/lis2dw12.c
 Do you see any difference between the above 2 drivers and our case or do these also need to change?

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.h:
```diff
@@ -125,8 +129,30 @@ enum fxos_trigger_type {
+
 struct fxos8700_config {
-	struct i2c_dt_spec i2c;
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+	const struct spi_dt_spec spi;
+#elif DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+	const struct i2c_dt_spec i2c;
+#endif
```
This can be a union

sumitbatra-nxp replied:
Changed as per your suggestion

MaureenHelm commented on drivers/sensor/fxas21002/fxas21002.c:
```diff
@@ -18,6 +18,129 @@ static const uint32_t sample_period[] = {
+	.byte_write = fxas21002_byte_write_spi,
+	.reg_field_update = fxas21002_reg_field_update_spi,
+};
+#elif DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
```
```suggestion
#endif
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
```

sumitbatra-nxp replied:
Acknowledged

MaureenHelm reviewed (changes requested):
Some of the conditional logic incorrectly assumes that all instances of the sensor will be on the same type of bus
```c
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
...
#elif DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
...
#endif
```

This needs to be changed to:
```c
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
...
#endif
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
...
#endif
```

bperseghetti committed:
boards: rddrone_fmuk66 Enable 3.3v regulator

Enables the 3.3v regulator lines on RDDRONE board.
Required to operate servo motors and I2C sensors.
Tested on RDDRONE board.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>
Co-authored-by: Landon Haugh <landon.haugh1@gmail.com>

bperseghetti committed:
boards: rddrone_fmuk66 fix pwm pinmux and servos

Fixes the channel 2 PWM pinmux adds servos and PWM LEDs.
Required to operate servo motors and status lighting.
Tested on RDDRONE board.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>
Co-authored-by: Landon Haugh <landon.haugh1@gmail.com>

bperseghetti committed:
boards: rddrone_fmuk66 fix I2C bus

Ensures I2C data line is open drain.
Tested on RDDRONE.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Landon Haugh <landon.haugh1@gmail.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

bperseghetti committed:
boards: rddrone_fmuk66 add I2C sensors

Adds I2C sensors for BMM150 and BME280.
Tested with BME280 and BMM150 sensors on RDDRONE.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Landon Haugh <landon.haugh1@gmail.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

bperseghetti committed:
boards: rddrone_fmuk66 dts: fxos8700 on SPI

Added support for fxos8700 sensor over SPI bus on
RDDRONE board and proper selection through dts.
Tested with fxos8700 sensor on RDDRONE.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

bperseghetti committed:
drivers: fxos8700 enablement for SPI and I2C

Added support for fxos8700 sensor over SPI and I2C.
Made the fxos driver APIs generic for I2C and SPI.
Tested with fxos8700 sensor on RDDRONE.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

bperseghetti committed:
boards: rddrone_fmuk66 dts: fxas21002 on SPI

Added support for fxas21002 sensor over SPI bus on
RDDRONE board and proper selection through dts.
Tested with fxas21002 sensor on RDDRONE.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

bperseghetti committed:
drivers: fxas21002 enablement for SPI and I2C

Added support for fxas21002 sensor over SPI and I2C.
Made the fxas driver APIs generic for I2C and SPI.
Tested with fxas21002 sensor on RDDRONE.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

bperseghetti reviewed (commented):
@danieldegrasse @MaureenHelm thank you all for the very beneficial suggestions, I believe that we now have it in an acceptable format.

sumitbatra-nxp commented:
@MaureenHelm @gmarull 
I just requested for re-review from both of you, as this page showed 2 changes requested from your side which we have already done. 
Kindly suggest us what needs to be done in order for these commits to go through to the open source.

MaureenHelm reviewed (changes requested):
Two minor requests, but otherwise LGTM. Thank you!

bperseghetti commented:
> Two minor requests, but otherwise LGTM. Thank you!

Awesome! Made the requested changes and rebased.

sumitbatra-nxp commented on drivers/sensor/fxas21002/fxas21002.c:
```diff
@@ -295,7 +473,7 @@ static const struct sensor_driver_api fxas21002_driver_api = {
 	};											\
 												\
-	SENSOR_DEVICE_DT_INST_DEFINE(inst, fxas21002_init, NULL,				\
+	DEVICE_DT_INST_DEFINE(inst, fxas21002_init, NULL,					\
```
Changed as per your request

sumitbatra-nxp commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -611,7 +767,7 @@ static const struct sensor_driver_api fxos8700_driver_api = {
 	static struct fxos8700_data fxos8700_data_##n;			\
 									\
-	SENSOR_DEVICE_DT_INST_DEFINE(n,					\
+	DEVICE_DT_INST_DEFINE(n,					\
```
Done as per your request

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -77,6 +77,35 @@
 		#phy-cells = <0>;
 	};
+
+	/* This regulator enables the 3V3_S line, which powers sensors on-board. */
+	reg-3v3-s {
+		compatible = "regulator-fixed";
+		regulator-name = "reg-3v3-s";
+		enable-gpios = <&gpiob 8 GPIO_ACTIVE_HIGH>;
+		startup-delay-us = <2000>;
+		regulator-always-on;
+		status = "okay";
+	};
+
+	/* This regulator enables the 3V3_S line, which powers sensors on-board. */
+	reg-3v3-neth {
+		compatible = "regulator-fixed";
+		regulator-name = "reg-3v3-neth";
+		enable-gpios = <&gpiob 3 GPIO_ACTIVE_LOW>;
+		regulator-always-on;
+		status = "okay";
+	};
+
+	/* This regulator enables the 3V3_S line, which powers sensors on-board. */
```
same comment for different regulators?

sumitbatra-nxp replied:
We have changed the comments for different regulators

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -106,6 +109,18 @@
 
+	servo: servo {
+		compatible = "pwm-servo";
+		pwms = <&ftm0 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH1 */
+			   <&ftm0 3 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH2 */
+			   <&ftm0 4 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH3 */
+			   <&ftm0 5 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH4 */
+			   <&ftm3 6 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH5 */
+			   <&ftm3 7 PWM_MSEC(20) PWM_POLARITY_NORMAL>; /* FMU_CH6 */
```
please fix alignment here, make sure your editor is properly configured.

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -106,6 +109,18 @@
 	};
 
+	servo: servo {
+		compatible = "pwm-servo";
```
where is this compatible found?

sumitbatra-nxp replied:
Servo motor is in the form of an app rather than a driver.
Please check out this README -
https://github.com/zephyrproject-rtos/zephyr/blob/7ad2e604bb3155c09a484717e1252c8f404640c0/samples/basic/servo_motor/README.rst

gmarull replied:
this compatible is not available outside of that sample.

sumitbatra-nxp replied:
@gmarull  Just to clarify.. Are you suggesting us to keep this servo node in the board's overlay file rather than in board dts ?
Like what's present for bbc_microbit in samples/basic/servo_motor/boards/bbc_microbit.overlay

gmarull replied:
I guess in your case you have a board that really has servo specific connectors, unlike bcc_microbit which is a general purpose board. So, it likely makes sense to move servo-motor compatible to `dts/bindings/...` folder if DT maintainers are fine with it.

sumitbatra-nxp replied:
@gmarull 
Only adding yaml binding for pwm-servo will be sufficient ? Or will you still need a driver like - drivers/led/led_pwm.c

gmarull replied:
bindings are sufficient, they are really the schema for DT content. It is not necessary to provide a driver for each DT compatible, in fact, some may never have one (e.g. the CPU). How DT hardware info is used may vary.

sumitbatra-nxp replied:
@gmarull .. Kindly check again and let us know if you are satisfied.

sumitbatra-nxp replied:
@gmarull .. The current servo_motor sample directory already has the pwm-servo.yaml present in -samples/basic/servo_motor/dts/bindings/pwm-servo.yaml
Now, If I add your proposed changes in the generic dts/bindings directory then servo_motor sample build fails. 

So, I think with due respect to the current servo_motor software design, we should add our board changes in the overlay file - *zephyr/samples/basic/servo_motor/boards/rddrone_fmuk66.overlay*

gmarull replied:
Yes, that is also a valid option, definitely.

sumitbatra-nxp replied:
@gmarull.. kindly re-review

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -310,6 +311,22 @@ zephyr_udc0: &usbotg {
 	pinctrl-0 = <&i2c1_default>;
 	pinctrl-names = "default";
+
+	/*
+	 * This board does not have a BME280, it has a BMP280.
+	 * The two parts are incredibly similar, so the BME280 driver should work.
+	 */
```
"should" or "works"? has this been tested properly?

sumitbatra-nxp replied:
@gmarull .. Yes it works and has been tested thoroughly, we will edit the comment

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -341,6 +342,26 @@ zephyr_udc0: &usbotg {
+	 * fxos8700  - CS#0
+	 * fxas21002 - CS#1
+	 */
+	cs-gpios = <&gpiob 10 GPIO_ACTIVE_LOW>,
+				<&gpiob 9 GPIO_ACTIVE_LOW>;
```
fix indentation here.

gmarull commented on dts/bindings/sensor/nxp,fxos8700-spi.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2018, NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: FXOS8700 6-axis accelerometer/magnetometer sensor
+
+compatible: "nxp,fxos8700"
+
+include: ["spi-device.yaml", "nxp,fxos8700-common.yaml"]
```
where are the changes adding support for SPI? If added later this breaks bisectability.

sumitbatra-nxp replied:
SPI changes have been added to the fxos8700 sensor driver, please -  drivers/sensor/fxos8700.c

gmarull commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -18,6 +18,132 @@ LOG_MODULE_REGISTER(FXOS8700, CONFIG_SENSOR_LOG_LEVEL);
 
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+#define DIR_READ(a)  ((a) & 0x7f)
+#define DIR_WRITE(a) ((a) | (1 << 7))
+#define ADDR_7(a) ((a) & (1 << 7))
```
`BIT(7)`

sumitbatra-nxp replied:
Done as suggested

gmarull commented on dts/bindings/sensor/nxp,fxas21002-common.yaml:
```diff
@@ -3,11 +3,18 @@
+      required: false
+      description: |
+        RST pin
+        This pin defaults to active high when consumed by the sensor.
```
"active high", but you set active low just above, is this correct?

sumitbatra-nxp replied:
Changed it to  - "This pin defaults to active low when consumed by the sensor"

gmarull commented on drivers/sensor/fxas21002/fxas21002.c:
```diff
@@ -18,6 +18,131 @@ static const uint32_t sample_period[] = {
 };
 
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+#define DIR_READ(a)		((a) | (1 << 7))
```
`BIT(7)`

gmarull commented on drivers/sensor/fxas21002/fxas21002.c:
```diff
@@ -18,6 +18,131 @@ static const uint32_t sample_period[] = {
+	uint8_t old_val;
+
+	if (fxas21002_byte_read_spi(dev, reg, &old_val) < 0) {
+		return -EIO;
```
Please propagate error codes, don't do conversions as you're loosing information.

MaureenHelm commented:
Please fix the failed compliance check

bperseghetti commented:
@MaureenHelm @danieldegrasse @gmarull Thank you all once again for the great reviews and suggestions on our PR. I hope that we have managed to fix all the remaining issues and suggestions.

bperseghetti reviewed (dismissed):
I think this all meets the requested changes now.

sumitbatra-nxp commented:
@gmarull @MaureenHelm : Kindly let us know if you require any further changes from our side..

dleach02 commented:
@gmarull can you revisit your review items?

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -342,6 +343,26 @@ zephyr_udc0: &usbotg {
+	 * fxas21002 - CS#1
+	 */
+	cs-gpios = <&gpiob 10 GPIO_ACTIVE_LOW>,
+		   <&gpiob 9 GPIO_ACTIVE_LOW>;
```
who uses CS1?

sumitbatra-nxp replied:
@gmarull .. Kindly check again and let us know if you are satisfied

gmarull replied:
missing fxas21002 entry below?

sumitbatra-nxp replied:
Added it back.. please check

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -363,6 +363,16 @@ zephyr_udc0: &usbotg {
 		status = "okay";
 	};
 
+	fxas21002@1 {
```
CS1 needs to be added in this commit

sumitbatra-nxp replied:
Ok, Acknowledged

sumitbatra-nxp replied:
@gmarull .. We have made your suggested changes.. kindly check again

gmarull commented on dts/bindings/sensor/nxp,fxas21002-i2c.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: FXAS21002 3-axis gyroscope sensor
+
+compatible: "nxp,fxas21002"
```
@galak @mbolivar-nordic, do we support defining 2 identical compatibles but with different properties? (see (*)) I see some other sensors doing that, so I guess they are merged altogether?

mbolivar-nordic replied:
Yes, this is fine. Note that including `i2c-device.yaml` below ensures that `on-bus: i2c` is in the final merged binding file. That combined with the `spi-device.yaml` include in the `nxp,fxas21002-spi.yaml` file creates two separate entries in the `devicetree.edtlib.EDT` object's `_compat2binding` map.

The purpose of this is precisely to allow for bus-specific properties for the same compatible.

This is documented in https://docs.zephyrproject.org/latest/build/dts/bindings.html#on-bus.

gmarull commented on dts/bindings/sensor/nxp,fxas21002-spi.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: FXAS21002 3-axis gyroscope sensor
+
+compatible: "nxp,fxas21002"
```
@galak @mbolivar-nordic (*)

gmarull commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -363,6 +363,16 @@ zephyr_udc0: &usbotg {
 	};
 
+	fxas21002@1 {
+		compatible = "nxp,fxas21002";
```
this breaks bisectability: you first need to make driver/bindings changes, then update boards.

sumitbatra-nxp replied:
@gmarull board and binding changes are part of the same commit and driver changes are in the next commit
So, shall we make driver changes as first commit and board and bindings as next?

gmarull replied:
Valid sequences could be: dts -> driver -> board, or dts+driver -> board

sumitbatra-nxp replied:
Acknowledged !

sumitbatra-nxp replied:
@gmarull .. We have made the change as per your suggestion.. Kindly re-review

bperseghetti commented:
@gmarull should be good to merge in now.

gmarull commented on samples/basic/servo_motor/boards/rddrone_fmuk66.overlay:
```diff
@@ -0,0 +1,14 @@
+/ {
+	servo: servo {
+		compatible = "pwm-servo";
+		pwms = <&ftm0 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH1 */
+		       <&ftm0 3 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH2 */
+		       <&ftm0 4 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH3 */
+		       <&ftm0 5 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH4 */
+		       <&ftm3 6 PWM_MSEC(20) PWM_POLARITY_NORMAL>, /* FMU_CH5 */
+		       <&ftm3 7 PWM_MSEC(20) PWM_POLARITY_NORMAL>; /* FMU_CH6 */
```
note: sample will just pick 1st entry, all others can't be tested with the sample as is. In fact, a servo can only be driven by a single PWM channel. You may comment other entries, but leave them in case people wants to try.

sumitbatra-nxp replied:
Done.. please check

gmarull reviewed (approved):
thanks for the continued effort

sumitbatra-nxp commented:
@MaureenHelm .. I think I mistakenly asked for a re-review from you, so can you please approve once again.

sumitbatra-nxp commented:
@gmarull .. thanks for helping us out. Really appreciated!
